### ISSUE TITLE
Switch to web.py mainline

### ISSUE THREAD

**Issue by anandology** [UCL-INGI/INGInious#117]
INGInious uses custom fork of web.py as it uses couple of customizations. IMHO, those customizations can be better achieved by extending the web.application class in  INGInious code, without having to maintain the fork.

There are two important customizations. 

1. Using dicts for URL mapping
2. Using objects instead of classes in the URL mapping

Using dicts for URL mapping is not a nice thing as it doesn't maintain the order and order is often very important. However, INGInious can still support it by simply extending the web.application.

    class Application(web.application):
        def init_mapping(self, mapping):
            self.mapping = WebPyCustomMapping(mapping)

INGInious uses objects because it want to make job_manager, user_manager etc. available to the page objects. That can also be achieved by keeping them in the application and access them via the application using properties. 

    class INGIniousPage(object):
        @property
        def app(self):
            return web.ctx.app_stack[0]

        @property
        def plugin_manager(self):
            return self.app.plugin_manager

        @property
        def course_factory(self):
            return self.app.course_factory

Or:

    class INGIniousPage(object):
        def __init__(self):
           app = web.ctx.app_stack[0]
           self.plugin_manager = app.plugin_manager
           self.course_factory = app.course_factory

In general, it doesn't look like a good idea to keep the life of a page object across requests. 

Also, the Application class looks like an ideal candidate for application initialization and cleanup code as well. That would simplify the code in `app.py` a lot.

If there are any more issues that are stoping you from using web.py mainline, I'll be happy to review them see if they can be integrated into web.py.

**Comment by GuillaumeDerval** [comment]
Thanks for the insights, that's greatly appreciated and we will integrate them in the (very) near future :-).

Once the ongoing refactor for python 3/zmq is done (end of the week), either @anthonygego or me will focus on this.

I completely agree the fork is not really needed, it was more a (bad) simple solution to solve (not so well) a particular problem one year ago.

Is the py3 branch from webpy available on pypi?

**Comment by anandology** [comment]
Py3 branch is not on pypi yet. I think it'll be released by end of this week.

**Comment by GuillaumeDerval** [comment]
I'm closing this as the need of the fork was dropped in 3cd71 ; The new_backend branch (with Py3) will be merged as soon as web.py with py3 support is available on PyPI :-)

I still have to implement your comments about initialization and cleanup with methods inside the app object, but it will be for another issue ;-)

**Comment by anandology** [comment]
Sorry, for the delay in pushing web.py py3 version to pypi. Will push the
dev version today so that ingenious can start using it.

**Comment by anandology** [comment]
web.py py3 version is available on pypi now. You can install it using:

    pip install web.py>=0.40.dev

**Comment by 31416r** [comment]
As we do not use the fork anymore, I suggest deleting the webpy-INGI repo altogether..

Also, Guillaume, I think it would be a good idea to create the issue about init/cleanup methods already.

**Comment by GuillaumeDerval** [comment]
The issue has already been created yesterday: #127

For webpy-INGI, I'm waiting the next release on Pypi to remove it/hide it.

**Comment by 31416r** [comment]
Sorry for the issue, I just noticed it after posting :)

Ok to wait for the next release to delete the repo. Have we a deadline for this ?

Also, I guess there will be some changes to the documentation for this release. Have we an issue for that ?

**Comment by GuillaumeDerval** [comment]
I'll first update our instance to the latest version and make sure it is stable enough.

We did not set a hard deadline yet, but I do not see how it could be after the end of this month.

**Comment by 31416r** [comment]
Nice phrasing for the deadline ;-)
